### PR TITLE
UniformsGroup: Support dynamic UBOs, booleans, and fixes 

### PR DESCRIPTION
**Pull request by RenaudRohlinger**
**Description:**

This PR introduces several improvements and fixes to the UniformsGroup:

**Dynamic UBOs support**: The UniformsGroup now supports dynamic Uniform Buffer Objects (UBOs). UniformsGroup now correctly handles boolean uniforms.

Bug fixes and improvements: This PR also includes several bug fixes and improvements related to UniformsGroup functionality, ensuring better stability and compatibility with different rendering pipelines.

**Changes:**

Added support for dynamic UBOs in UniformsGroup
Implemented proper handling of boolean uniforms in UniformsGroup
Fixed various bugs and issues related to UniformsGroup functionality
Updated documentation and examples to showcase the new features and fixes

**Testing:**

The changes have been tested on various platforms and browsers for a bunch of months. I'm currently backporting to upstream our fork of threejs but any additional reviews and feedback are welcome. Also, updated the ubo example to demonstrate the new features and improvements in action.

image

### CODE REVIEW

**Comment by Mugen87** [comment]
It seems there is a merge conflict in `webgl2_ubo`.

**Comment by Mugen87** [comment]
The E2E screenshot definitely needs an update since the visuals of the example have changed.

**Comment by RenaudRohlinger** [comment]
what's the command that needs to be used to resolve this issue?
I'm using `npm run make-screenshot webgl2_ubo.html` to generate the image which seems to fail.

**Comment by LeviPesin** [comment]
I think you should try rebasing and only then making screenshot.

**Comment by RenaudRohlinger** [comment]
Indeed! Fixed my issue by merging with upstream 👍

**Comment by Mugen87** [comment]
The builds should definitely not be included in the PR. 

I'm not sure why the build fails. Let me download your branch so I can make a test on my computer.

File: examples/webgl2_ubo.html [hunk 1]
```diff
@@ -183,23 +220,56 @@
...
+
+			lightingUniformsGroup.add( new THREE.Uniform( [ new THREE.Vector4( Math.random() * 40 - 20, 1, Math.random() * 40 - 20, 0 ), new THREE.Vector4( col[ 0 ], col[ 40 ], col[ 2 ], 0 ) ] ) ); // light position
 
 			init();
+
 			animate();
 
+			function dispose() {
+			
+				scene?.traverse( function ( object ) {
```
Comment (Mugen87) [inline examples/webgl2_ubo.html:239]: Please do not use optional chaining. Only ECMA2018 is allowed.

File: src/core/UniformsGroup.js [hunk 2]
```diff
@@ -5,7 +5,7 @@ let id = 0;
 
 class UniformsGroup extends EventDispatcher {
 
-	constructor() {
+	constructor( count ) {
```
Comment (Mugen87) [inline src/core/UniformsGroup.js:8]: What is the purpose of the new `count` parameter of `UniformsGroup`?

**Comment by Mugen87** [comment]
TBH, I have a bit of a hard time to review this PR because it has so many changes at once.

I think it's better if you split the PR into multiple ones. E.g. fixes in one PR, add support for booleans in another one. And support for dynamic UBOs last?

**Comment by RenaudRohlinger** [comment]
Ok I will split and create multiple PRs this later this week

**Comment by Mugen87** [comment]
@RenaudRohlinger Thanks!

**Pull request closed without merging**